Implement Data Downloader for BioCypher

ElnatanSamuel opened the pull request:
- **Update for Issue #35** 

### Folder Structure
> downloader/
> ├── __init__.py
> ├── download_data.py
> ├── download_manager.py
> ├── protocols/
>  .   ├── init__.py
>  .   ├── base.py
>  .   └── http.py
> 
> config/
> ├── download.yaml

### Changes Introduced
The new DownloadManager class and associated scripts provide a flexible framework for downloading various data sources required by the project. Key features and changes include:

- DownloadManager Class: A new class that manages the downloading of data from multiple sources, supporting HTTP and HTTPS

- Retry Logic: Implemented retry logic with exponential backoff for HTTP requests to handle transient network issues.

- Error Handling: Enhanced error handling to log detailed error messages and continue processing other downloads even if one fails.

- Helper Methods: Added modular helper methods to handle different types of downloads, including single URL and multiple URLs

-  Progress Bars: Integrated progress bars using tqdm to provide users with visual feedback on download progress.

- Configuration Files: Utilized YAML configuration files to specify download sources, making it easy to add or modify data sources without changing the code.

- Special Cases Handling: Addressed special cases such as non-SSL verification for certain sources and unpacking archives post-download.

### Example usage

**To download all files**

> **python downloader/download_data.py --output-dir data/** 

**To download specific adapter files**

> **python downloader/download_data.py --output-dir data/ --source genconde**

dawit-melka reviewed (changes requested):
Could you update the links for the adapters I commented? and please add the Example usage you mention in the pr description in the ReadMe file as well.
Thanks!

ElnatanSamuel commented:
> Could you update the links for the adapters I commented? and please add the Example usage you mention in the pr description in the ReadMe file as well. Thanks!

I have changed the requested url's and added the description to the readme file.

dawit-melka reviewed (approved):
Looks good!
